share/janitor/tasks.py

```diff
@@ -2,19 +2,40 @@
 
 import celery
 
+from django.db.models import Exists
+from django.db.models import OuterRef
+
 from share import tasks
 from share.models import RawDatum
+from share.models import NormalizedData
 
 
 logger = logging.getLogger(__name__)
 
 
 @celery.shared_task(bind=True)
-def rawdata_janitor(self, limit=100):
+def rawdata_janitor(self, limit=500):
     """Find RawDatum that do not have a NormalizedData process them
     """
     count = 0
-    for rd in RawDatum.objects.select_related('suid__source_config').filter(normalizeddata__isnull=True).order_by('id')[:limit].iterator():
+
+    # NOTE: Do NOT use .iterator here. It will create a temporary table and eat disk space like no other
+    # the limit lets this query fit nicely in memory and actually finish executing
+    # Be very careful about changing this query. If you do change it, make sure the EXPLAIN looks something like this:
+    # Limit  (cost=1.13..Much Smaller Numbers)
+    #   ->  Nested Loop  (cost=1.13..Big Numbers)
+    #         Join Filter: (share_sourceuniqueidentifier.source_config_id = share_sourceconfig.id)
+    #         ->  Nested Loop  (cost=1.13..Big Numbers)
+    #               ->  Nested Loop Anti Join  (cost=0.56..Big Numbers)
+    #                     ->  Seq Scan on share_rawdatum  (cost=0.00..Big Numbers)
+    #                     ->  Index Only Scan using share_normalizeddata_c0e72696 on share_normalizeddata  (cost=0.56..Small Numbers)
+    #                           Index Cond: (raw_id = share_rawdatum.id)
+
+    qs = RawDatum.objects.select_related('suid__source_config').annotate(
```

Comment:
You're right but I think I would prefer the ability to turn logging on rather than have this be slightly faster. The additional joins don't add all that much overhead, it's the anti-join (and the order by) that brought the query to a crawl.